doc/source/fq_nmod_sparse_mat.rst

```diff
@@ -0,0 +1,295 @@
+    corresponding ``vals``, all of length ``nnz``, assumes sorted by rows 
+    with no duplicate (row, col) indices
+
+.. function:: void fq_nmod_sparse_mat_append_col(fq_nmod_sparse_mat_t M, const fq_nmod_struct *v, const fq_nmod_ctx_t ctx)
+
+    Add a dense column to the right of the matrix
+
+.. function:: void fq_nmod_sparse_mat_append_row(fq_nmod_sparse_mat_t M, const fq_nmod_sparse_vec_t v, const fq_nmod_ctx_t ctx)
+
+    Add a sparse row to the bottom of the matrix
+
+
+Conversion to/from dense matrix
+--------------------------------------------------------------------------------
+
+.. function:: void fq_nmod_sparse_mat_from_dense(fq_nmod_sparse_mat_t M, const fq_nmod_mat_t dM, const fq_nmod_ctx_t ctx)
+
+    Converts the dense matrix ``dM`` to a sparse matrix ``M``
+
+.. function:: void fq_nmod_sparse_mat_to_dense(fq_nmod_mat_t dM, const fq_nmod_sparse_mat_t M, const fq_nmod_ctx_t ctx)
+
+    Converts the sparse matrix ``M`` to a dense matrix ``dM``
+
+Windows, concatenation, and splitting
+--------------------------------------------------------------------------------
+
+.. function:: void fq_nmod_sparse_mat_window_init(fq_nmod_sparse_mat_t window, const fq_nmod_sparse_mat_t M, slong r1, slong c1, slong r2, slong c2, const fq_nmod_ctx_t ctx)
+
+    Constructs a window on the sparse matrix ``M`` between rows ``r1`` and ``r2`` 
+    and cols ``c1`` and ``c2`` (valid as long as original matrix remains unmodified)
+
+.. function:: void fq_nmod_sparse_mat_window_clear(fq_nmod_sparse_mat_t window, const fq_nmod_ctx_t ctx)
+
+    Clears a window
+
+.. function:: void fq_nmod_sparse_mat_concat_horizontal(fq_nmod_sparse_mat_t B, const fq_nmod_sparse_mat_t M1, const fq_nmod_sparse_mat_t M2, const fq_nmod_ctx_t ctx)
+
+    Horizontally concatenates two matrices ``M1`` and ``M2`` into block matrix ``B``
+
+.. function:: void fq_nmod_sparse_mat_concat_vertical(fq_nmod_sparse_mat_t B, const fq_nmod_sparse_mat_t M1, const fq_nmod_sparse_mat_t M2, const fq_nmod_ctx_t ctx)
+
+    Vertically concatenates two matrices ``M1`` and ``M2`` into block matrix ``B``
+
+.. function:: void fq_nmod_sparse_mat_split_horizontal(fq_nmod_sparse_mat_t M1, fq_nmod_sparse_mat_t M2, const fq_nmod_sparse_mat_t B, slong c, const fq_nmod_ctx_t ctx)
+
+    Splits ``B`` horizontally into two submatrices ``M1`` and ``M2``, dividing at column ``c``
+
+.. function:: void fq_nmod_sparse_mat_split_vertical(fq_nmod_sparse_mat_t M1, fq_nmod_sparse_mat_t M2, const fq_nmod_sparse_mat_t B, slong r, const fq_nmod_ctx_t ctx)
+
+    Splits ``B`` vertically into two submatrices ``M1`` and ``M2``, dividing at row ``r``
+
+
+Permutation
+--------------------------------------------------------------------------------
+
+.. function:: void fq_nmod_sparse_mat_permute_cols(fq_nmod_sparse_mat_t M, slong *Q, const fq_nmod_ctx_t ctx)
+
+    Permutes the columns indices of ``M`` according to ``Q``, and re-sorts each row
```

Comment:
You have to specify what Q is. Is it the set of old columns indices in their new order, or the index that each of the old columns is placed in. I also don't understand what re-sorting rows is all about.